### ISSUE TITLE
search return same   result even with different images using dataloader with batch-

### ISSUE THREAD

**Issue by flydragon2018** [facebookresearch/faiss#1103]
# Summary
 I build one vector dataset by extract feature from images using pre-trained model.
 when I process the image one by one, it works.

 however,  when I process using dataloader with batch and num_worker ,
 after I build index and save it to disk. when I  do search using the index built by batch-processing,
 It always return same result, obviously it is wrong. however the index built from image one by one works perfectly.    

I have compares the way I convert pickle file to vectors, the result have same size. 

what is wrong?   It takes too long time to process big dataset without batch-processing.

# Platform
 

OS: 
ubuntu18.04
Faiss version: 
1.6.1 
Faiss compilation options: 
using cuda and python3
Running on: GPU
- [ ] CPU
- [ ] GPU

Interface:  PYTHON3
- [ ] C++
- [ ] Python

# Reproduction instructions
  dataloader  with batch=1 , in contrast with batch /worker= 8/16/ etc.

**Comment by flydragon2018** [comment]
I have compared the feature-vector extracted by one and by batch.

it seems that the orders of image fetching by one and by batch are not same. 

however,  that should not cause the  faiss.search  return same result.

any idea?

**Comment by mdouze** [comment]
There is no notion of a dataloader in Faiss, so I assume it is from pytorch.
It is unclear what the problem is here. Can you post a minimal reproduction script?

**Comment by flydragon2018** [comment]
Thank you for your reply.

as I said earlier,   I use VGG/RESNET to extract image feature vectors  for FAISS to build index.
Yes,  It is dataloader of pytorch.

when I load the image by dataloader with batchsize=1, num_workers=0,   put the feature vectors 
in a list,  dump it to PICKLE.   load the pickle file in FAISS  and then build index.  use that index to 
query image,  It return right answer.

however,  when I load image in dataloader with batchsize=8,num_workers=8(or any other pairs),
because every time I got  [batchsize-label, [vector,vector, ...]),  I  append the vectors into list by iteration, and dump to PICKLE file.  in the same way,  I load the pickle file for FAISS and build index 
corresponding.  the later index does return  search result.  however obviously it is wrong.
every time , the search result is same with different images. 

 the code is not at hand.  I may post minimal script later.

**Comment by flydragon2018** [comment]
import .....
def extractor():
     device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    ## extract features
    print("extracting feature:" )
    model_ft = models.resnet18(pretrained=True)

    ### strip the last layer
    feature_extractor = torch.nn.Sequential(*list(model_ft.children())[:-1])
    
    testdata_transform= transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        ])

    testdata_dir = 'car_data'
    
    bs=2
    nw=2

    image_testdatasets =datasets.ImageFolder(root="car_data",transform=testdata_transform)
                   
    testdataloaders = torch.utils.data.DataLoader(image_testdatasets, batch_size=bs,
                                              shuffle=False, num_workers=nw)
               
    
    print("test image data:")
    print(testdataloaders)

    searchList=[]
    resultList=[]
    labelList=[]
    count=0
    j=0
    for i, (inputs, labels) in enumerate(testdataloaders):
        ids=labels.numpy()
        ## add labels to list
        n=len(labels)
        for k in range(n):
            labelList.append([j+k,labels])
        j+=n
             
        output = feature_extractor(inputs) # output now has the features corresponding to input x
        
        

        np_vec=output.detach().numpy()
        print("np_vec shape :")
        print(np_vec.shape)
        
        
        m=np_vec.shape[0]

        for v in range(m):
            print("np_vec[{}].shape".format(v))
            print(np_vec[v].shape)
            resultList.append(np_vec[v].reshape(-1,512,1,1)) 
            
        print(resultList[count+v].shape)
        count+=m

       
    print("total count=",count)
    
 
    # # List of numbers
    # L = resultList
    # # Open the file in write binary mode
    with open('featureVectorsbybatch.pickle', 'wb') as F:
         # Dump the list to file
         pickle.dump(resultList, F)
    
    with open('labelListbybatch.pickle', 'wb') as F:
         # Dump the list to file
         pickle.dump(labelList, F)
   
    return resultList

**********************************************************************************************
above star line is my batch-process code.
the one-by-one process with batchsize=1,num_workers=0.  and there is no loop iteration for the
returned output since batchsize=1.

**Comment by mdouze** [comment]
I don't see any Faiss in the code above.

**Comment by flydragon2018** [comment]
there is nothing much related to FAISS  search  but the feature vectors Pickle file extracted.
because if I used the Pickle file dumped by one-by-one processing, It will return right result.
so the problem relying on the parallel dataloader and feature vectors serialization.


**************************************************************
import faiss
import numpy as np

import resnet18onefeature as rnf 

import time


def search_img_faiss(indexfile=None, imgVector=None,topN=3):

    print(imgVector)

    # read in faiss index first:
    #load index
    cpu_index=faiss.read_index(indexfile)
    res = faiss.StandardGpuResources()
    gpu_index=faiss.index_cpu_to_gpu(res, 0, cpu_index)

    print("add vectors to index")
    #search
    D,I=gpu_index.search(imgVector, topN)

    return  D,I 
s=time.time()
nq,xq=rnf.extractor()

xq=np.array(xq).reshape(1,512)

xq=np.array(xq).reshape(1,512)

d,i=search_img_faiss(indexfile="./my_index/myindexbybatch",imgVector=xq,topN=3)

print(i)
print(d)
t=time.time()
print("\n\n total search time in seconds:",t-s)

**Comment by flydragon2018** [comment]
perhaps buildindex file :
by the way, the batchsize used in building index is same as used in feature extracting.

************************************************************

import faiss
import numpy as np
import os

from load_img_vectors import load_img_vectors_bybatch

index_path = "./my_index"
ids_vectors_path = './my_ids_vectors'

# ---------------------  Search
TOP_N = 5
SIMILARITY = 5



# prepare index, the dimension of feature extracted from resnet18 is 512 
dim = 512
index = faiss.IndexFlatL2(dim) 

# we need only a StandardGpuResources per GPU
res = faiss.StandardGpuResources()
flat_config = faiss.GpuIndexFlatConfig()
flat_config.device = 1

index = faiss.GpuIndexFlatL2(res, dim, flat_config)


id = 0
index_dict = {}

n,vecs=load_img_vectors_bybatch("featureVectorsbybatch.pickle")

batchsize=16

print("n=",n)
print("vecs[0]=",vecs[0])

xb = np.random.random((n*batchsize, dim)).astype('float32')

print("vecs length:",n)
 

print("xb shape:", xb.shape)

i=0
j=0
k=0

for item in vecs:
    
    print(item.shape)
    print("item[0] shape")
    print(item[0].shape)
    j=len(item)
    
    #xb[k:l]=item.reshape(j,dim)
    for t in range(j):
        #print(k+t,t)
        xb[k+t] =item[t].reshape(dim)
      
    print("xb[k+t,:].shape ")
    print(xb[k+t,:].shape)

    k+=j
    i+=1
    
index.add(xb)
# save index, convert index from gpu to cpu first and then save it.
index_cpu = faiss.index_gpu_to_cpu(index)
faiss.write_index(index_cpu, os.path.join(index_path,"myindexbybatch"))

**Comment by mdouze** [comment]
Sorry, still can't repro.

**Comment by mdouze** [comment]
no activity, closing.
